[WIP] Run tests as part of the CI process

AdelekeBankole opened the pull request:
On moving forward for the test suite in CAM-ML. This PR aims to run our tests as part of the continuous integration process.

AdelekeBankole committed:
add ci job for UbuntuTestSuite

AdelekeBankole committed:
add workflow run

AdelekeBankole committed:
add steps needed for job

AdelekeBankole committed:
add pFUnit to steps required in job

jatkinson1000 commented on .github/workflows/test_suite_ubuntu.yml:
```diff
@@ -0,0 +1,72 @@
+  # Triggers the workflow on pushes to open pull requests with code changes
+  pull_request:
+    paths:
+      - '.github/workflows/test_suite_ubuntu.yml'
+      - '**.c'
+      - '**.cpp'
+      - '**.fypp'
+      - '**.f90'
+      - '**.F90'
+      - '**.pf'
+      - '**.py'
+      - '**.sh'
+      - '**CMakeLists.txt'
+      - '**requirements.txt'
+      - '**data/*'
```
We do not have a number of these filetypes, or do not use them for the Fortran tests so can remove.
```suggestion
      - '.github/workflows/test_suite_ubuntu.yml'
      - '**.f90'
      - '**.F90'
      - '**CMakeLists.txt'
      - '**requirements.txt'
```

jatkinson1000 commented on .github/workflows/test_suite_ubuntu.yml:
```diff
@@ -0,0 +1,72 @@
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        std: ["f2008", "f2018"]
```
We do not need to check against Fortran standards

```suggestion
```

jatkinson1000 commented on .github/workflows/test_suite_ubuntu.yml:
```diff
@@ -0,0 +1,72 @@
+        uses: actions/setup-python@v5
+        with:
+          python-version: '3.x'
+
+      # MPI is required by pFUnit
+      - name: Install an MPI distribution
+        run: |
+          sudo apt update
+          sudo apt install mpich
+
+      - name: Install pFUnit
+        run: |
+          export FC=/usr/bin/gfortran
+          export MPIF90=/usr/bin/mpif90
+          git clone -b v4.10.0 https://github.com/Goddard-Fortran-Ecosystem/pFUnit.git
+          mkdir pFUnit/build
+          cd pFUnit/build
+          cmake ..
+          make -j 4 install    
```
Out test suite does not require pFUnit or MPI
```suggestion
```

jatkinson1000 reviewed (changes requested):
Hi @AdelekeBankole This is a good start and it is good to see it is running with GitHub CI!

There are a few things that I don't think we need here (I know I said to start with the FTorch CI as an example) that I have provided 'suggestions' to remove.

The next step I think is to define a series of `steps` to run the test suite as we would on a local machine. This will include:

- [x] Install any python dependencies
- [x] Generate the sounding files using python
- [x] Build the code using CMake
- [x] Run the test suites
- [x] Modify the test suites to return exit codes for success/failure

Let me know if you want me to take another look after further changes or have any queries.

AdelekeBankole committed:
Update .github/workflows/test_suite_ubuntu.yml

Co-authored-by: Jack Atkinson <109271713+jatkinson1000@users.noreply.github.com>

AdelekeBankole committed:
Update .github/workflows/test_suite_ubuntu.yml

Co-authored-by: Jack Atkinson <109271713+jatkinson1000@users.noreply.github.com>

AdelekeBankole committed:
Update .github/workflows/test_suite_ubuntu.yml

Co-authored-by: Jack Atkinson <109271713+jatkinson1000@users.noreply.github.com>

AdelekeBankole committed:
add steps to run test suite

jatkinson1000 commented on .github/workflows/test_suite_ubuntu.yml:
```diff
@@ -0,0 +1,76 @@
+          python3 -m pip install -r requirements.txt
+
+      # Generate sounding files
+      - name: Generate sounding files 
+        run: |
+          python -m venv venv
+          source venv/bin/activate
+          python sounding_to_netcdf.py
+          deactivate
```
```suggestion
      - name: Generate sounding files 
        run: |
          cd YOG_convection/resources/
          python3 -m venv venv
          source venv/bin/activate
          python3 -m pip install -r requirements.txt
          python3 sounding_to_netcdf.py
          deactivate
          cd ../../
```

jatkinson1000 reviewed (commented):
I think the CI currently fails because it does not build the sounding in the correct location.
Try the suggestion below, and we can take a closer look this afternoon.

AdelekeBankole committed:
Update .github/workflows/test_suite_ubuntu.yml

Co-authored-by: Jack Atkinson <109271713+jatkinson1000@users.noreply.github.com>

AdelekeBankole committed:
add python linter, ruff

jatkinson1000 commented:
@AdelekeBankole Looking at the latest failure on the test script it seems python netcdf cannot find the HDF5 library or something.
It may be that you need to install netcdf4 and/or HDF5 libraries to the CI runner as well.

AdelekeBankole committed:
add hdf5 in requirements.txt

AdelekeBankole committed:
remove hdf5in requirements.txt

AdelekeBankole committed:
add and install hdf5

AdelekeBankole committed:
allow root privileges in getting hdf5

AdelekeBankole committed:
add and install netcdf lib to dependencies

jatkinson1000 commented:
@AdelekeBankole Looks like the versions are too strict in the YOG/resources/requirements.txt
Perhaps remove the specifiers and see if things run OK.

Though note you may need to restrict numpy to <2.0 as I recall it had issues with the major version change and I do not know if they have been resolved yet.

AdelekeBankole commented:
> @AdelekeBankole Looks like the versions are too strict in the YOG/resources/requirements.txt Perhaps remove the specifiers and see if things run OK.
> 
> Though note you may need to restrict numpy to <2.0 as I recall it had issues with the major version change and I do not know if they have been resolved yet.

@jatkinson1000, you mean restricting it to `>=3.7,<3.11` ?

AdelekeBankole committed:
restrict python version

jatkinson1000 committed:
remove pins from dependencies in sounding generation requirements.

jatkinson1000 committed:
remove ruff from testing suite as covered in the linting checks CI.

jatkinson1000 committed:
test changing python version.

jatkinson1000 committed:
Add SAM sounding file.

It is a binary, but unlikely to change and cannot be generated in the CI due to issues
with netcdf save in python requiring os.getlogin which doesn't work on
CI runners for some reason (similar issues can be found elsewhere).

jatkinson1000 committed:
remove sounding generation from CI workflow as sounding has been uploaded as NetCDF file.

jatkinson1000 committed:
add netcdf-Fortran to the dependencies installed in the testing CI.

AdelekeBankole committed:
update python run options

jatkinson1000 commented:
@AdelekeBankole I meant from the pip dependencies in the requirements.txt file - the workflow was complaining that it could not find the exact version of numpy.

See PR #90 to this branch that has a patch to do this.
This worked, but there were still issues (documented various places online) with running netcdf.save on GitHub CI so I uploaded the NetCDF file as an artifact to fix this.

I also added netCDF-Fortran that was missing from the dependencies.

My branch now builds the test suite so hopefully allows you to continue troubleshooting.

jatkinson1000 committed:
run test from test directory.

jatkinson1000 committed:
Fix ruff errors and pin version to prevent unanticipated issues from version updates in future.

jatkinson1000 committed:
Minor tidying of the python CI workflow file.

jatkinson1000 committed:
remove duplicate running of CAM interface test script in test-suite.

Co-authored-by: Marion <56403724+MarionBWeinzierl@users.noreply.github.com>

AdelekeBankole committed:
Merge pull request #90 from m2lines/jwa-ci-dependencies-patch

Fix various problems for building and running tests in GitHub CI

AdelekeBankole committed:
add exit codes run step

AdelekeBankole committed:
add stop 999

AdelekeBankole committed:
try to break code

AdelekeBankole committed:
remove exit code

jatkinson1000 reviewed (commented):
That's looking good @AdelekeBankole !
Now you just need to un-break it ;)

A quick thought - did you also want to run the other test suite as well?

AdelekeBankole committed:
remove the break

AdelekeBankole committed:
add YOG convection test in ci

AdelekeBankole committed:
remove YOG convection test in ci

jatkinson1000 reviewed (approved):
Hi @AdelekeBankole This all looks like good work to me and I have approved.

I have made one small suggestions regarding what I believe is an unnecessary python setup in the job that you might want to look at.
I'm happy for you to merge after checking this.

AdelekeBankole committed:
Update .github/workflows/test_suite_ubuntu.yml

Co-authored-by: Jack Atkinson <109271713+jatkinson1000@users.noreply.github.com>
